src/avram/postgres_url.cr

```diff
@@ -1,42 +1,82 @@
 class Avram::PostgresURL
-  getter database, username, password, hostname, port
+  getter database, username, password, hostname, port, query
+  getter url : String = ""
 
   def initialize(
     @database : String,
-    @hostname : String,
+    @hostname : String = "",
     @username : String = "",
     @password : String = "",
-    @port : String = ""
+    @port : String = "",
+    @query : String = ""
```

Comment:
Cold we make these nilable instead of empty strings? I wonder if that'd clarify some of the code

Reply:
Yeah, we can do that 👍

Reply:
Just realized this might make the code a little more complicated... Right now I'm stripping spaces from all of the components, but making them nilable means I have to check if there's a value before stripping on all of them

```crystal
@username = username.strip
io << username unless username.empty?

# becomes

if username && !username.empty?
  @username = username.strip

  io << username
end
```

hmm.. maybe that's fine? 🤔  After writing that, it might be clearer even though it's more lines.

Reply:
How set are you on having these nilable? 😅  I'm fighting the compiler now..

```crystal
if username
  @username = username.to_s.strip

  # even though the compiler knows it's not nil here, the getter is still nilable
  io << URI.encode_www_form(username) unless username.not_nil!.empty?
end
```

There's a few things that get cleaned up, but a few other things that become a lot more difficult. Also, the type of port becomes `Int32?` now because that's what `URI.parse` will return.